### ISSUE TITLE
Question: SnapshotCollector

### ISSUE THREAD

**Issue by yahorsi** [microsoft/ApplicationInsights-dotnet#1353]
Classic ASP.NET
Do we need to do anything extra to make SnapshotCollector working? Or, in other words, should SnapshotCollector.SnapshotCollectorTelemetryProcessor be explicitly added in the ApplicationInsights.config or adding corresponding nuget package is enought? If yes, could you please provide sample.

Core ASP.NET
What should be done to make snapshots collection working in the ASP.NET Core application? 

PS: Sorry for spam, trying to make AI working 3rd day (( The docs are not great (

**Comment by uncleyo** [comment]
So I've just spent almost a day trying to make snapshot debugger work in a legacy ASP.NET application and now it finally works!
While there's no guarantee that the issue I've encountered is the same as yours was, I believe it is quite possible, because the symptoms I experienced were basically the same.

**TL;DR: it is `<assemblyBinding>`.**

### Problem

Application Insights probably tries to be resilient to failures by preferring to (silently?) ignore them.

And that is what it does to `SnapshotCollectorTelemetryProcessor`. It simply silently skips it if it fails to load and proceeds happily without it.
You can verify this by entering any garbage instead of `Microsoft.ApplicationInsights.SnapshotCollector.SnapshotCollectorTelemetryProcessor, Microsoft.ApplicationInsights.SnapshotCollector`
as `Type` attribute value in `ApplicationInsights.config`.

And if you check `TelemetryConfiguration.Active.DefaultTelemetrySink.TelemetryProcessors`, it will simply miss those it failed to load.

A sure way to verify that problem is with assembly loading, just add this line somewhere in startup code:
```csharp
var _ = typeof(SnapshotCollectorTelemetryProcessor);
```
That should trigger the error.

AFAIK, unlike with class library projects, `Web.config` in ASP.NET applications cannot benefit from `<AutoGenerateBindingRedirects>`, while NuGet package manager in VS won't do it either once you move to `PackageReference` (or to "new csproj with globs" like I did). So you have to maintain it somewhat manually.

### "Fix"

As of today, assuming you have these reference versions in your `csproj` file,
```xml
<PackageReference Include="Microsoft.ApplicationInsights.SnapshotCollector" Version="1.3.7" />
<PackageReference Include="Microsoft.ApplicationInsights.Web" Version="2.14.0" />
```
you will probably need at least these redirects in Web.config
```xml
<assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
  <dependentAssembly>
    <assemblyIdentity name="Microsoft.ApplicationInsights" publicKeyToken="31bf3856ad364e35" culture="neutral" />
    <bindingRedirect oldVersion="0.0.0.0-2.14.0.17971" newVersion="2.14.0.17971" />
  </dependentAssembly>
  <dependentAssembly>
    <assemblyIdentity name="Microsoft.ApplicationInsights.SnapshotCollector" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
    <bindingRedirect oldVersion="0.0.0.0-1.3.7.0" newVersion="1.3.7.0" />
  </dependentAssembly>
  <dependentAssembly>
    <assemblyIdentity name="System.Diagnostics.DiagnosticSource" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
    <bindingRedirect oldVersion="0.0.0.0-4.0.4.0" newVersion="4.0.4.0" />
  </dependentAssembly>
</assemblyBinding>
```

After adding those, I could see the `SnapshotUploader` directory created on disk and soon after simulating exceptions, some snapshots in the portal.

### Extra quirks

As described above, the configuration by default is wrongly added to the root in `ApplicationInsights.config` and not under default sink.

Additionally, that sink configuration created by `Microsoft.ApplicationInsights.Web` contains duplicate entries for `AdaptiveSamplingTelemetryProcessor`, which will probably cause much less telemetry to be sampled than expected in some cases.

**Comment by yahorsi** [comment]
Tried on the latest VS and latest snapshots collector package and it still fails to install itself correctly. The AI config after package installation is corrupted

**Comment by yahorsi** [comment]
@uncleyo Could you please share your AI and packages configs?

Thanks!

**Comment by uncleyo** [comment]
I haven't installed it into my _existing_ solution.
What I did is I created an "empty" ASP.NET MVC app from template, commited that to git and then added package via package manager to that, so I can see the diff. I then simply copied the resulting ApplicationInsights.config to my actual project and added the PackageReference's.

And yes, my actual solution is a "Frankenstein", which is a classic ASP.NET app with a .net core SDK based csproj (obviously targeting full framework). But that doesn't matter. As long as you get your `assemblyBinding`s right, it should work.

I don't know what you mean when you say "it fails to install itself". Does nuget package manager fail to install package or what?

**Comment by yahorsi** [comment]
>>I don't know what you mean when you say "it fails to install itself". Does nuget package manager fail to install package or what?
It's just an old bug that was not fixed yet. If you add SnapshotsCollector package it tries to modify ApplicationInsights.config but does it incorrectly that results to the broken config.

I've spent today whole day trying to make AI work for the classic app, and yes I've made simple empty app, add there AI, pushed and it does not work (( No events in the AI while I see them locally in the VS

**Comment by RoyDoron** [comment]
Could you have the snapshot debugger option along with "Failure" and "Performance" Telemetry?

Objective: Enable Snapshot debugger to Diagnose session file (*.diagsession) for a specific timeframe hooked to an app service (basically debug prod)

Current State: We have Production application Insights with "Failures" and "Performance" Section Enabled - which mean we can view live telemetry.

Problem: We have experience prod exceptions which the current telemetry is unable to provide proper insights as to why they happens nor the root cause of them.

Problem with the implementation of the debugger (Microsoft Official Response): We cannot migrate the app to a workspace based without loosing the data available for Failures" and "Performance" Sections. Which means, we cannot have both solutions. Its either Debugger or Performance/Failures Metrices.

Looking for Insights or suggestions on a workaround or permanent solution to triage prod issues using the debugger or without, cause Azure AppInsights is so frustrating to use

(EFS stack implementation is scheduled for 2022, but until than, we need to manage)

Cheers!

References:

https://docs.microsoft.com/en-us/azure/azure-monitor/app/create-workspace-resource
https://docs.microsoft.com/en-us/azure/azure-monitor/app/snapshot-debugger-vm?toc=/azure/azure-monitor/toc.json
https://docs.microsoft.com/en-us/azure/azure-monitor/app/snapshot-debugger

**Comment by pharring** [comment]
I need to close this stale issue, but I see there are some unresolved comments above.
@RoyDoron @yahorsi @uncleyo, if you still need help with Snapshot Collector, please open a new issue at https://github.com/microsoft/ApplicationInsights-SnapshotCollector/issues/new/choose
and we'll try and address it there.
